### PR TITLE
Fix Grappler ArgMax/ArgMin wrong results for saturating ops in float32

### LINKED ISSUE #118675: Grappler `ArgMax`/`ArgMin` monotonic stripping is wrong for strictly-monotonic ops with float32 saturation/overflow
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

TensorFlow 2.22.0-dev20260429

### Custom code

Yes

### OS platform and distribution

_No response_

### Mobile device

_No response_

### Python version

_No response_

### Bazel version

_No response_

### GCC/compiler version

_No response_

### CUDA/cuDNN version

_No response_

### GPU model and memory

_No response_

## Summary

Grappler's `ArithmeticOptimizer` transforms `ArgMax(f(x))` into `ArgMax(x)` for ops classified as strictly monotonic (tanh, exp, sigmoid). While these functions are mathematically strictly monotonic, float32 representation limits cause them to produce **tied** outputs (saturation to 1.0, overflow to inf, underflow to 0), changing the argmax result.

This is distinct from [#116436](https://github.com/tensorflow/tensorflow/issues/116436), which covers non-strictly-monotonic ops (relu, floor, ceil). That issue explicitly states: *"The remaining ops (Exp, Log, Sigmoid, Tanh, Softplus, etc.) are strictly monotonic and the optimization is valid for them."* This bug shows the optimization is also invalid for those ops when float32 limits create ties.

## Scope

Affects all strictly-monotonic ops that saturate or overflow in float32:

| Op | Mechanism | Example | Eager | Grappler |
|---|---|---|---|---|
| `tanh` | saturation to ±1.0 | `argmax(tanh([10, 20, 5]))` | **0** | 1 |
| `exp` | overflow to inf | `argmax(exp([100, 200, 50]))` | **0** | 1 |
| `sigmoid` | saturation to 1.0 | `argmax(sigmoid([20, 40, 5]))` | **0** | 1 |
| `exp` | underflow to 0 | `argmin(exp([-100, -200, -50]))` | **0** | 1 |
| `tanh` | saturation to -1.0 | `argmin(tanh([-5, -10, -20]))` | **1** | 2 |

The pattern is consistent: when two or more inputs produce the same float32 output (ties), Grappler's optimization returns the wrong index because it computes argmax on the original values (no ties) instead of the transformed values (with ties).

Both argmax and argmin are affected.

## Practical Impact

Temperature-scaled classification: `argmax(sigmoid(logits/T))` where logits have large magnitude or T is small. The sigmoid saturates, creating ties, and Grappler silently returns the wrong class.

Feature selection: `argmax(tanh(features))` where features have large magnitude. Common when features are not normalized.

## Root Cause

In `tensorflow/core/grappler/op_types.cc`, `IsElementWiseMonotonic()` returns `true` for tanh, exp, sigmoid, etc. The `OptimizeMaxOrMinOfMonotonic` pass in `ArithmeticOptimizer` uses this to strip the op from argmax/argmin. The transformation `ArgMax(f(x)) → ArgMax(x)` requires that `f(a) > f(b) ⟺ a > b`, which holds for strictly monotonic functions over the reals, but **not** in float32 when saturation/overflow maps distinct values to the same float32 output.

The fix in [#116709](https://github.com/tensorflow/tensorflow/pull/116709) for #116436 only addressed non-strictly-monotonic ops (relu, floor, etc.) and explicitly kept tanh/exp/sigmoid/etc. as "valid" for stripping.

## Suggested Fix

The optimization `ArgMax(f(x)) → ArgMax(x)` is only safe when `f` is injective over the representable float range. For ops that saturate (tanh, sigmoid) or overflow (exp), the optimization should be removed or restricted to inputs known to be within the injective range.

Concretely, remove `Tanh`, `Sigmoid`, `Exp`, and any other saturating/overflowing op from the list of ops eligible for argmax/argmin stripping in `IsElementWiseMonotonic()`.

## Environment

- TensorFlow 2.22.0-dev20260429
- Grappler `arithmetic_optimization` is enabled by default
- Eager mode: correct
- `arithmetic_optimization=False`: correct
- Affects both CPU and GPU (Grappler is device-independent)

### Standalone code to reproduce the issue

```shell
import tensorflow as tf

x = tf.constant([[10.0, 20.0, 5.0]])
# tanh(x) = [1.0, 1.0, 0.9999] → argmax = 0 (first 1.0, tie-breaking)
# x       = [10, 20, 5]         → argmax = 1 (20 is largest)

print("eager:", tf.math.argmax(tf.math.tanh(x), axis=-1).numpy())  # [0]

tf.config.optimizer.set_experimental_options({"arithmetic_optimization": True})
print("Grappler:", tf.function(
    lambda x: tf.math.argmax(tf.math.tanh(x), axis=-1))(x).numpy())  # [1] WRONG


Observed on TensorFlow 2.22.0-dev20260429.
```

### Relevant log output

```shell

```

### PR DESCRIPTION
**Pull request by Cyrax321**
## Problem

Grappler's `ArithmeticOptimizer` transforms `ArgMax(f(x)) → ArgMax(x)` for ops listed in `IsElementWiseMonotonic()`. This optimization is only valid when `f` is **strictly injective** over float32 — i.e., `f(a) == f(b)` implies `a == b`. Several ops in the list violate this due to float32 saturation and overflow, producing silently wrong results.

Reported in issue #118675.

```python
@tf.function
def f(x): return tf.argmax(tf.math.tanh(x))
x = tf.constant([10.0, 20.0, 5.0])
print(tf.argmax(tf.math.tanh(x)))  # eager:   0  ✓ (first 1.0)
print(f(x))                        # grappler: 1  ✗ (wrong: argmax of raw [10,20,5])
```

## Root cause

In `tensorflow/core/grappler/op_types.cc`, `IsElementWiseMonotonic()` listed ops that are mathematically injective over the reals but not injective over float32:

| Op | Float32 failure | Tied output |
|---|---|---|
| Tanh | Saturates to ±1.0 | `tanh([10, 20]) → [1.0, 1.0]` |
| Exp | Overflows to +inf | `exp([100, 200]) → [inf, inf]` |
| Sigmoid | Saturates to 0 and 1.0 | `sigmoid([20, 40]) → [1.0, 1.0]` |
| Erf | Saturates to ±1.0 | `erf([4, 8]) → [1.0, 1.0]` |
| Elu / Selu | Saturates to −1.0 | `elu([-20, -100]) → [-1.0, -1.0]` |
| Relu | Maps negatives to 0.0 | `relu([-2, -1]) → [0.0, 0.0]` |
| Relu6 | Saturates at 6.0 | `relu6([7, 100]) → [6.0, 6.0]` |
| Floor / Ceil / Rint | Many-to-one | `floor([1.1, 1.9]) → [1.0, 1.0]` |
| Sign | Many-to-one | `sign([2, 5]) → [1.0, 1.0]` |
| Atanh | Undefined outside (−1, 1) | |
| Expm1 | Overflows to +inf | Like Exp |

## Fix

Remove the non-injective ops from `kMonotonicNonDecreasingOps`. The following ops are kept (strictly injective over all finite float32 values):

`Acosh`, `Asin`, `Asinh`, `Atan`, `Log`, `Log1p`, `Sinh`, `Softsign`, `Softplus`, `Sqrt`

Non-increasing: `Acos`, `Erfc`, `Neg`, `Rsqrt` (unchanged)

## Tests

Added 8 tests to `arithmetic_optimizer_test.cc`:

- 7 tests verifying that ArgMax/ArgMin of non-injective ops (Tanh, Exp, Sigmoid, Erf, Elu, Relu6, Relu) are **not** rewritten by Grappler
- 1 test verifying that ArgMax(Sinh(x)) is still optimized (Sinh is safe)

Fixes #118675

### CODE REVIEW

File: tensorflow/python/ops/linalg_grad.py [hunk 1]
```diff
@@ -374,14 +374,46 @@ def _GetGradWrt(output_grad, other_operand, input_shape, input_subs,
...
+  cumprod_right = array_ops.concat(
+      [math_ops.cumprod(s[..., 1:], axis=-1, reverse=True), ones], axis=-1)
+  prod_except = cumprod_left * cumprod_right  # [..., n]
+  # Sign correction: det(A) = det(U) * prod(s) * det(V); s >= 0 so sign is
+  # carried by the orthogonal factors U and V.
+  sign_uv = gen_linalg_ops.matrix_determinant(
+      u) * gen_linalg_ops.matrix_determinant(v)  # [...]
+  # adj(A)^T = sign_uv * U @ diag(prod_except) @ V^T
+  a_adj_t = sign_uv[..., None, None] * math_ops.matmul(
+      u * prod_except[..., None, :], v, transpose_b=True)
```
[the remark this answers is not in the document]
Comment (Cyrax321) [reply tensorflow/python/ops/linalg_grad.py]: Thank you for the review. The complex matrix fix is tracked in PR #122823 (the  gradient PR). That PR has already been updated with  and .

Note: the formula suggested here ( + ) was tested numerically and produces incorrect results for complex matrices. The correct formula is:

```python
sign_uv = det(U) * conj(det(V))
a_adj_t = sign_uv * (U * prod_except) @ V^H   # adjoint_b=True
```

This was verified against the exact formula `grad = det(A) * inv(A, adjoint=True)` and matches to atol=1e-10. The test `testDetGradComplexInvertible` added in #122823 covers this case.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Thank you for addressing this silent numerical issue in Grappler's graph optimizations. Finding and fixing these subtle mathematical flaws is extremely valuable.

However, after reviewing the proposed changes, there are several critical issues and architectural concerns that need to be addressed before this can be merged.

## Critical Issues

### 1. Regression on `Max`, `Min`, and `MaxPool` Optimizations
The optimization `Max(f(x)) -> f(Max(x))` (and similarly for `Min` and `MaxPool`) is mathematically valid for **all** monotonic functions, regardless of whether they are injective or saturating.
Proof: If $f$ is non-decreasing, then $x_i \le x_{\max} \implies f(x_i) \le f(x_{\max})$. Thus, $\max_i f(x_i) = f(\max_i x_i)$.

By removing saturating ops (like `Tanh`, `Exp`, `Floor`) from `IsElementWiseMonotonic`, this PR accidentally **disables valid optimizations** for `Max`, `Min`, and `MaxPool` reductions. This is a regression in optimization coverage.

### 2. Logical Error: `Relu` is NOT Strictly Injective
The PR description claims that `Relu` is kept because it is "strictly injective over all finite float32 values". This is mathematically incorrect.
`Relu(x) = max(0, x)`. For all negative inputs $(x < 0)$, `Relu` saturates to `0.0`.
For example, `Relu(-1.0) == 0.0` and `Relu(-2.0) == 0.0`. It is a many-to-one mapping in this domain.

Keeping `Relu` in the optimization list for `ArgMax`/`ArgMin` will lead to silent bugs.
Consider $x = [-2.0, -1.0]$:
- Eager/Correct: `ArgMax(Relu(x)) = ArgMax([0.0, 0.0]) = 0` (first max).
- Optimized (if rewritten): `ArgMax(x) = 1` (since -1.0 > -2.0).
The result differs, yielding corrupted indices.

### 3. Floating Point Precision Limits and Injectivity
Many other ops listed as "kept" (e.g., `Softsign`, `Atan`, `Sinh`) are practically non-injective in `float32` due to saturation or overflow at limits:
- `Softsign(x) = x / (1 + |x|)` approaches `1.0`. For $x > 8.3 \times 10^6$, `1 + x == x` in float32, so `Softsign` becomes exactly `1.0`, creating ties.
- `Atan(x)` approaches $\pi/2$. For large finite values, it becomes exactly $\pi/2$ in float32.
- `Sinh(x)` overflows to `+inf` for finite $x > 88.7$. Multiple large finite values will yield `+inf`, losing injectivity.

## Suggested Approach

To fix `ArgMax`/`ArgMin` without regressing `Max`/`Min` optimizations:

1. **Keep `IsElementWiseMonotonic` intact** to preserve valid optimizations for `Max`, `Min`, and `MaxPool`.
2. In `arithmetic_optimizer.cc` (specifically in `OptimizeMaxOrMinOfMonotonicStage`), separate the logic for `ArgMax`/`ArgMin`.
3. Introduce a stricter classifier (e.g., `IsElementWiseStrictlyInjective`) only for `ArgMax`/`ArgMin` optimizations.
4. Ensure `Relu`, `Softplus`, and ops that overflow/saturate are explicitly excluded from this strictly injective set.

## Verification Unit Test

Here is a unit test that demonstrates how keeping `Relu` in this optimization chain leads to failures (it will fail if `Relu` is optimized to `ArgMax(x)`):

```cpp
TEST_F(ArithmeticOptimizerTest, OptimizeArgMaxOfReluFailure) {
  tensorflow::Scope s = tensorflow::Scope::NewRootScope();
  // Input with negative values where Relu saturates to 0.
  const auto x = ops::Const(s.WithOpName("x"), {-2.0f, -1.0f}, {1, 2});
  Output relu = ops::Relu(s.WithOpName("relu"), x);
  Output arg_max = ops::ArgMax(s.WithOpName("arg_max"), relu, 1);
  Output final_out = ops::Identity(s.WithOpName("final_out"), arg_max);

  GrapplerItem item;
  item.fetch = {"final_out"};
  TF_CHECK_OK(s.ToGraphDef(&item.graph));
  const auto tensors_expected = EvaluateNodes(item.graph, item.fetch);
  ASSERT_EQ(tensors_expected.size(), 1);

  GraphDef output;
  ArithmeticOptimizer optimizer;
  EnableOnlyOptimizeMaxOrMinOfMonotonic(&optimizer);
  OptimizeAndPrune(&optimizer, &item, &output);
  const auto tensors = EvaluateNodes(output, item.fetch);
  ASSERT_EQ(tensors.size(), 1);

  // Expected: ArgMax(Relu([-2, -1])) = ArgMax([0, 0]) = 0
  // Actual (if optimized): ArgMax([-2, -1]) = 1
  // This expectation should FAIL if the optimization is applied.
  test::ExpectTensorEqual<int64_t>(tensors[0], tensors_expected[0]);
}
```

**Comment by Cyrax321** [comment]
Thank you for the review. The points raised were correct, including the Relu case; that was an error in my original claim.

The change has been reworked as suggested:

  - `IsElementWiseMonotonic` is restored to its original op lists, so the Max/Min/MaxPool optimizations are unaffected.
  - Added a new `IsElementWiseStrictlyInjective` classifier in `op_types.cc`, used only for ArgMax/ArgMin in `OptimizeMaxOrMinOfMonotonicStage`. Value-producing reductions keep the plain monotonicity check.
  - The strict set is now `Acosh, Asin, Asinh, Log, Log1p, Neg, Rsqrt, Sqrt`. Relu, Softplus, Softsign, Atan, Sinh and the other saturating or overflowing ops are excluded, with the float32 reason for each documented in a comment. Acos is also excluded: it collapses all |x| below ~1e-8 to exactly acos(0), since acos(x) is approximately pi/2 - x near zero and pi/2 has an ulp of ~1.2e-7.
  - Tests: the Relu case is covered by `ArgMaxOfNonInjectiveOpIsNotOptimized_Relu` (same inputs, {-2, -1}). The previous Sinh positive test was incorrect and is now a negative test with overflowing inputs. Added `MaxOfSaturatingMonotonicOpIsStillOptimized_Tanh` to verify the Max path still rewrites through saturating monotonic ops.

Please let me know if anything else needs adjusting.

**Review by dmiltr3 — CHANGES_REQUESTED** [review]
Some of the tests added in `arithmetic_optimizer_test.cc` can be brittle in certain testing environments due to a subtle issue with test helpers and graph node ordering.

## Issue with `OptimizeAndPrune` and `VerifyGraphsMatch`

In tests where you expect the graph to **not** be optimized (e.g., `ArgMaxOfSaturatingOpIsNotOptimized_*`), calling `OptimizeAndPrune` followed by `VerifyGraphsMatch` can lead to spurious mismatches.

1.  `VerifyGraphsMatch` is strictly order-sensitive and compares nodes index-by-index.
2.  `ModelPruner` (called inside `OptimizeAndPrune`) may reorder nodes in the `GraphDef` even if nothing was actually pruned, causing the comparison to fail against the original graph.

## Recommended Fix

For tests verifying that **no optimization happened** (NOOP):

1.  Avoid using `OptimizeAndPrune`.
2.  Instead, call `optimizer.Optimize` exactly once. This keeps the original graph intact in `item.graph` and gives you the optimized graph in `output`.
3.  Compare `item.graph` (Original) directly with `output` (Optimized) using `VerifyGraphsMatch`. This bypasses `ModelPruner` entirely. Since the optimizer is expected to be a no-op, node ordering is invariant and the comparison is robust.

Here is the suggested pattern for your "NotOptimized" tests:

```cpp
  GraphDef output;
  ArithmeticOptimizer optimizer;
  EnableOnlyOptimizeMaxOrMinOfMonotonic(&optimizer);
  
  // Single pass to preserve original graph in item.graph and avoid ModelPruner reordering
  TF_EXPECT_OK(optimizer.Optimize(nullptr, item, &output));

  VerifyGraphsMatch(item.graph, output, __LINE__);
```

This pattern is more robust and avoids potential false positives where an optimization might fire but get masked by helpers verifying idempotency rather than original equality.

## Recommended Updates to Tests

We recommend applying this single-pass verification pattern to the following tests in your PR:
- `ArgMaxOfSaturatingOpIsNotOptimized_Tanh`
- `ArgMaxOfSaturatingOpIsNotOptimized_Sigmoid`
- `ArgMaxOfSaturatingOpIsNotOptimized_Erf`
- `ArgMinOfSaturatingOpIsNotOptimized_Elu`
- `ArgMaxOfSaturatingOpIsNotOptimized_Relu6`
- `ArgMaxOfSaturatingOpIsNotOptimized_Sinh`
- `ArgMaxOfNonInjectiveOpIsNotOptimized_Relu`
- `ArgMaxOfSaturatingOpIsNotOptimized_Exp`

**Comment by Cyrax321** [comment]
Thank you for the review. The single-pass pattern has been applied to all eight tests listed: each now calls `optimizer.Optimize` exactly once, leaving `item.graph` untouched, and compares the result directly against the
original graph, avoiding `OptimizeAndPrune` and the `ModelPruner` reordering entirely.

One small adjustment was needed while applying the pattern: `ArithmeticOptimizer::Optimize` itself performs a topological sort on the graph before running any rewrite stage, so node ordering can change even when the pass is a complete no-op. With the scope-generated graphs in these tests, the ArgMax dimension constant moves ahead of the unary op, and the index-by-index comparison in `VerifyGraphsMatch` then reports a spurious mismatch. The comparison therefore uses `CompareGraphs` (from `GrapplerTest`) instead. It sorts nodes by name before comparing, is already used for the same single-pass no-op pattern elsewhere in this file, and additionally verifies attributes and devices. The reordering was confirmed in isolation with a non-monotonic op (`Sin`) where the stage cannot fire: node inputs remain identical but the ordering still differs.

Happy to switch to a different comparison approach if preferred. Please let me know if anything else needs adjusting.

**Review by dmiltr3 — APPROVED** [review]

**Review by dmiltr3 — APPROVED** [review]
